browsermt/marian-dev
# Import matrix-multiply (PrepareA) from a separate wasm module

### Description

This PR resolves first 4 tasks of the issue https://github.com/browsermt/bergamot-translator/issues/204

List of changes:
- [x] Restructure [marian-dev](https://github.com/browsermt/marian-dev), allowing it to import different implementations of  [int8PrepareA](https://github.com/browsermt/marian-dev/blob/master/src/tensors/cpu/wasm_intgemm_interface.h#L148) for wasm target
- [x] Add a fallback (a polyfill) dummy implementation of [int8PrepareA](https://github.com/browsermt/marian-dev/blob/master/src/tensors/cpu/wasm_intgemm_interface.h#L148) (`src/tensors/cpu/wasm_intgemm_fallback.cpp` file) in [marian-dev](https://github.com/browsermt/marian-dev) for wasm target and export it from marian wasm module 
    - The fallback dummy implementation aborts initially
- [x] Change emscripten generated JS glue code to plug in this fallback implementation (until an optimized implementation is not landed in Firefox)

Please note that for integer 16-bit matrix multiplication, the wasm target still uses intgemm (we are not aborting for integer 16-bit matrix multiplication).

Added dependencies: none

### How to test
Follow the [README](https://github.com/browsermt/marian-dev/tree/master/wasm#-performing-translation-using-wasm-marian-decoder) instructions to run the wasm test page and observe the abort messages in browser console logs.

### Checklist
